cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/EOGWorklist.kt

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright (c) 2023, Fraunhofer AISEC. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ *                    $$$$$$\  $$$$$$$\   $$$$$$\
+ *                   $$  __$$\ $$  __$$\ $$  __$$\
+ *                   $$ /  \__|$$ |  $$ |$$ /  \__|
+ *                   $$ |      $$$$$$$  |$$ |$$$$\
+ *                   $$ |      $$  ____/ $$ |\_$$ |
+ *                   $$ |  $$\ $$ |      $$ |  $$ |
+ *                   \$$$$$   |$$ |      \$$$$$   |
+ *                    \______/ \__|       \______/
+ *
+ */
+package de.fraunhofer.aisec.cpg.helpers
+
+import de.fraunhofer.aisec.cpg.graph.Node
+import de.fraunhofer.aisec.cpg.graph.edge.PropertyEdge
+import java.util.IdentityHashMap
+
+/**
+ * A complete lattice is an ordered structure of values of type [T]. [T] could be anything, e.g., a
+ * set, a new data structure (like a range), or anything else. [T] depends on the analysis and
+ * typically has to abstract the value for the specific purpose.
+ *
+ * This class is actually used to hold individual instances of the lattice's elements and to compute
+ * bigger elements depending on these two elements.
+ *
+ * Implementations of this class have to implement the comparator, the least upper bound of two
+ * lattices.
+ */
+abstract class Lattice<T>(open val elements: T) : Comparable<Lattice<T>?> {
+    /**
+     * Computes the least upper bound of this lattice and [other]. It returns a new object and does
+     * not modify either of the objects.
+     */
+    abstract fun lub(other: Lattice<T>?): Lattice<T>
```

Comment:
Why nullable? Seems to introduce a lot of unnecessary nulls checks.

Reply:
You should be able to get rid of them. Makes the code much easier to read and maintain.

```patch
diff --git a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/EOGWorklist.kt b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/EOGWorklist.kt
index 7083dd7fc..dbe7ba5c5 100644
--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/EOGWorklist.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/EOGWorklist.kt
@@ -40,12 +40,12 @@ import java.util.IdentityHashMap
  * Implementations of this class have to implement the comparator, the least upper bound of two
  * lattices.
  */
-abstract class Lattice<T>(open val elements: T) : Comparable<Lattice<T>?> {
+abstract class Lattice<T>(open val elements: T) : Comparable<Lattice<T>> {
     /**
      * Computes the least upper bound of this lattice and [other]. It returns a new object and does
      * not modify either of the objects.
      */
-    abstract fun lub(other: Lattice<T>?): Lattice<T>
+    abstract fun lub(other: Lattice<T>): Lattice<T>
 
     /** Duplicates the object, i.e., makes a deep copy. */
     abstract fun duplicate(): Lattice<T>
@@ -55,14 +55,14 @@ abstract class Lattice<T>(open val elements: T) : Comparable<Lattice<T>?> {
  * Implements the [Lattice] over a set of nodes. The lattice itself is constructed by the powerset.
  */
 class PowersetLattice(override val elements: Set<Node>) : Lattice<Set<Node>>(elements) {
-    override fun lub(other: Lattice<Set<Node>>?) =
-        PowersetLattice((other?.elements ?: setOf()).union(this.elements))
+    override fun lub(other: Lattice<Set<Node>>) =
+        PowersetLattice(other.elements.union(this.elements))
 
     override fun duplicate() = PowersetLattice(this.elements.toSet())
 
-    override fun compareTo(other: Lattice<Set<Node>>?): Int {
-        return if (this.elements.containsAll(other?.elements ?: setOf())) {
-            if (this.elements.size > (other?.elements?.size ?: 0)) 1 else 0
+    override fun compareTo(other: Lattice<Set<Node>>): Int {
+        return if (this.elements.containsAll(other.elements)) {
+            if (this.elements.size > (other.elements.size)) 1 else 0
         } else {
             -1
         }
@@ -99,7 +99,8 @@ open class State<K, V> : IdentityHashMap<K, Lattice<V>>() {
     open fun needsUpdate(other: State<K, V>): Boolean {
         var update = false
         for ((node, newLattice) in other) {
-            update = update || node !in this || newLattice > this[node]
+            val current = this[node]
+            update = update || current == null || newLattice > current
         }
         return update
     }
@@ -122,13 +123,15 @@ open class State<K, V> : IdentityHashMap<K, Lattice<V>>() {
         if (newLattice == null) {
             return false
         }
-        if (newNode in this && this[newNode]?.let { it >= newLattice } == true) {
+
+        val current = this[newNode]
+        if (current != null && current >= newLattice) {
             // newLattice is "smaller" than the currently stored one. We don't add it anything.
             return false
-        } else if (newNode in this) {
+        } else if (current != null) {
             // newLattice is "bigger" than the currently stored one. We update it to the least
             // upper bound
-            this[newNode] = newLattice.lub(this[newNode])
+            this[newNode] = newLattice.lub(current)
         } else {
             this[newNode] = newLattice.duplicate()
         }
@@ -211,7 +214,7 @@ class Worklist<K : Any, N : Any, V>() {
     fun hasAlreadySeen(currentNode: K) = currentNode in alreadySeen
 
     /** Computes the meet over paths for all the states in [globalState]. */
-    fun mop(): State<N, V>? {
+    fun mop(): State<N, V> {
         val firstKey = globalState.keys.firstOrNull()
         val state = globalState[firstKey]
         for ((_, v) in globalState) {
@@ -236,7 +239,7 @@ inline fun <reified K : Node, V> iterateEOG(
     startNode: K,
     startState: State<K, V>,
     transformation: (K, State<K, V>, Worklist<K, K, V>) -> State<K, V>
-): State<K, V>? {
+): State<K, V> {
     val worklist = Worklist(mutableMapOf(Pair(startNode, startState)))
     worklist.push(startNode, startState)
 
diff --git a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/ControlDependenceGraphPass.kt b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/ControlDependenceGraphPass.kt
index 15a972ea7..094a283bf 100644
--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/ControlDependenceGraphPass.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/ControlDependenceGraphPass.kt
@@ -224,9 +224,8 @@ private fun <T : Node> PropertyEdge<T>.isConditionalBranch(): Boolean {
  */
 class PrevEOGLattice(override val elements: Map<Node, Set<Node>>) :
     Lattice<Map<Node, Set<Node>>>(elements) {
-    override fun lub(other: Lattice<Map<Node, Set<Node>>>?): Lattice<Map<Node, Set<Node>>> {
-        val newMap =
-            (other?.elements ?: mapOf()).mapValues { (_, v) -> v.toMutableSet() }.toMutableMap()
+    override fun lub(other: Lattice<Map<Node, Set<Node>>>): Lattice<Map<Node, Set<Node>>> {
+        val newMap = other.elements.mapValues { (_, v) -> v.toMutableSet() }.toMutableMap()
         for ((key, value) in this.elements) {
             newMap.computeIfAbsent(key, ::mutableSetOf).addAll(value)
         }
@@ -235,16 +234,14 @@ class PrevEOGLattice(override val elements: Map<Node, Set<Node>>) :
 
     override fun duplicate() = PrevEOGLattice(this.elements.toMap())
 
-    override fun compareTo(other: Lattice<Map<Node, Set<Node>>>?): Int {
+    override fun compareTo(other: Lattice<Map<Node, Set<Node>>>): Int {
         return if (
-            this.elements.keys.containsAll(other?.elements?.keys ?: setOf()) &&
-                this.elements.all { (k, v) -> v.containsAll(other?.elements?.get(k) ?: setOf()) }
+            this.elements.keys.containsAll(other.elements.keys) &&
+                this.elements.all { (k, v) -> v.containsAll(other.elements[k] ?: setOf()) }
         ) {
             if (
-                this.elements.keys.size > (other?.elements?.keys?.size ?: 0) ||
-                    this.elements.any { (k, v) ->
-                        v.size > (other?.elements?.get(k) ?: setOf()).size
-                    }
+                this.elements.keys.size > (other.elements.keys.size) ||
+                    this.elements.any { (k, v) -> v.size > (other.elements.get(k) ?: setOf()).size }
             )
                 1
             else 0
```